# Statistical and Spatial Analysis of the Apartment Segment in the Spanish Real Estate Market (2019) - Extraction and Preparation

## Introduction

In this project, the focus will be on analyzing the apartment segment within the real estate market. Apartments are a particularly relevant type of housing in urban and suburban areas, where the demand for accessible and functional options is high. This analysis aims to provide a detailed understanding of the characteristics, trends, and factors that influence the price and availability of apartments in different regions. By using exploratory data analysis (EDA) techniques, key patterns and important correlations will be extracted to provide valuable insights for buyers, investors, and other stakeholders in the real estate sector.

Although the main focus is on the analysis of the apartment segment, the data transformation section will be conducted on the entire dataset. This approach ensures a clean and prepared DataFrame ready for future analyses involving different types of properties and market segments.

## Objectives

The primary goal of this project is to perform a statistical analysis of apartments and family houses for sale in various Spanish provinces during 2019. The analysis will employ geolocation techniques and data visualization tools such as Tableau. The aim is to provide a detailed understanding of the characteristics and spatial distribution of the Spanish housing market during this period. Given the target audience of U.S. investors and private buyers, terminology will be adapted to align with their context. The study aims to achieve the following objectives:

- Clean and explore the data to ensure accuracy and reliability.
- Use geolocation techniques to convert property addresses into geographic coordinates.
- Calculate and summarize key statistical measures for property prices and characteristics.
- Visualize the geographic distribution and identify regional price differences.
- Analyze correlations between property features and their prices.
- Apply clustering techniques to identify patterns and segment the real estate market.
- Perform hypothesis testing to compare prices across different regions and assess the impact of specific features.
- Develop and validate linear regression models to predict property prices.
- Design and build an interactive Tableau dashboard for dynamic data exploration.
- Present key findings from the data analysis and provide practical recommendations based on the insights gained.


## Data Source and Structure

The dataset used in this project was sourced from Kaggle: [Spanish Housing Dataset](https://www.kaggle.com/datasets/thedevastator/spanish-housing-dataset-location-size-price-and/data)

Was originally collected through web scraping from the [Idealista S.A.U.](https://www.idealista.com/) by github user [trueuoc](https://github.com/trueuoc) and published under the license: [CC0 1.0 Universal (CC0 1.0) - Public Domain Dedication](https://creativecommons.org/publicdomain/zero/1.0/legalcode.en)

It contains key variables that describe real estate listings, including property details (e.g., price, size, condition), location (e.g., city, district), and amenities (e.g., air conditioning, garden, pool). The data is structured with clear identifiers and includes both categorical and numerical variables, allowing for robust analysis.

### Variables

- **ad_description**: Descripcion de la propiedad.
- **ad_last_update**: Ultima actualización de la información.
- **air_conditioner**: Whether the property has air conditioning (0: No, 1: Yes).
- **balcony**: Whether the property has a balcony (0: No, 1: Yes).
- **bath_num**: Number of bathrooms in the property.
- **built_in_wardrobe**: Whether the property has built-in wardrobes (0: No, 1: Yes).
- **chimney**: Whether the property has a chimney (0: No, 1: Yes).
- **condition**: Condition of the property (e.g., second-hand/good condition).
- **construct_date**: Year of construction of the property.
- **energetic_certif**: Property’s energy certification.
- **floor**: Floor on which the property is located.
- **garage**: Indicates whether it has a garage space.
- **garden**: Whether the property has a garden (0: No, 1: Yes).
- **ground_size**: Size of the property's land.
- **heating**: Whether the property has heating (0: No, 1: Yes).
- **house_id**: A unique property ID number.
- **house_type**: Type of housing.
- **kitchen**: Whether the property has a kitchen.
- **lift**: Whether the property has an elevator (0: No, 1: Yes).
- **loc_city**: City where the property is located.
- **loc_district**: District.
- **loc_full**: Full address of the property.
- **loc_neigh**: Neighborhood.
- **loc_street**: Street.
- **loc_zone**: Zone.
- **m2_real**: Actual square meters of the property.
- **m2_useful**: Usable square meters.
- **obtention_date**: Date when the data was collected.
- **orientation**: Property's orientation.
- **price**: Property price.
- **reduce_mobility**: Whether the house is adapted for people with reduced mobility.
- **room_num**: Number of rooms in the property.
- **storage_room**: Whether the property has a storage room (0: No, 1: Yes).
- **swimming_pool**: Whether the property has a swimming pool (0: No, 1: Yes).
- **terrace**: Whether the property has a terrace (0: No, 1: Yes).
- **unfurnished**: Whether the property is unfurnished.
- **number_of_companies_prov**: Number of companies in the province.
- **population_prov**: Population of the province.
- **companies_prov_vs_national_%**: Percentage of companies in the province compared to the national total.
- **population_prov_vs_national_%**: Percentage of the population in the province compared to the national total.
- **renta_media_prov**: Average income in the province.

## Data Collection and Preprocessing

### Import libraries

In [1]:
# Modulos
import pandas as pd
import gc


**Pandas Settings**

In [10]:
# Set Pandas to Display All Columns
pd.set_option('display.max_columns', None)
pd.options.display.float_format = '{:.2f}'.format

#### scripts.functions documentation

**Funtions:**

- **filterdf**(df, col1, val1, col2, val2): Generates a dataframe filtered by two variables.
- **binary_categorical_view**(dataframe): Visualizes binary categorical variables: `air_conditioner`, `chimney`, `garden`, `storage_room`, `swimming_pool`, `terrace`.
- **categorical_features_view**(dataframe): Visualizes categorical variables: `room_num`, `bath_num`, `condition`.
- **boxplot_view**(dataframe, column): Visualizes a specific column.
- **boxplot_view_wo**(dataframe, column): Visualizes a specific column without displaying outliers.
- **distribution_views**(dataframe): Visualizes histograms for the variables: `m2_real`, `price`.
- **boxplot_with_mean**(dataframe, group_col, target_col, show_outliers, figsize): Displays boxplots of a variable grouped by another variable's values.
- **plot_histogram**(df, column, bins=20, kde=True, figsize=(10, 6), xlim=None): Displays a histogram of a variable, which can be segmented.
- **plot_distribution_by_price_segment**(df): Groups data by price segment and type of property, generating a heatmap.

### Data Import and Initial Exploration

In [11]:
raw_data = pd.read_csv('data/source/spanish_houses.csv')

In [16]:
print(raw_data.head(1).T)

                                                                               0
ad_description                 Precio chalet individual en la localidad de Ab...
ad_last_update                                Anuncio actualizado el 27 de marzo
air_conditioner                                                                0
balcony                                                                        0
bath_num                                                                       2
built_in_wardrobe                                                              0
chimney                                                                        0
condition                                               segunda mano/buen estado
construct_date                                                               NaN
energetic_certif                                                             NaN
floor                                                                  2 plantas
garage                      

In [7]:
raw_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 41 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   ad_description                 95426 non-null   object 
 1   ad_last_update                 100000 non-null  object 
 2   air_conditioner                100000 non-null  object 
 3   balcony                        100000 non-null  object 
 4   bath_num                       100000 non-null  object 
 5   built_in_wardrobe              100000 non-null  object 
 6   chimney                        100000 non-null  object 
 7   condition                      86059 non-null   object 
 8   construct_date                 32059 non-null   object 
 9   energetic_certif               74691 non-null   object 
 10  floor                          79693 non-null   object 
 11  garage                         40811 non-null   object 
 12  garden                         

**Dataset Summary**

- The dataset consists of **100,000 entries** and **41 columns**.
- The data types include `object`, `float64`, and `int64`.
- Several columns contain missing values.
- The dataset occupies **31.3 MB** in memory, making it manageable for analysis.
- This initial exploration highlights the need for data cleaning, especially in columns with a high proportion of missing values.

## Data Transformation

### Data Filtering
Remove the variables that will not be used in this analysis.

In [8]:
raw_data.drop(columns=['ad_description', 
                       'ad_last_update', 
                       'house_id', 
                       'obtention_date', 
                       'unfurnished',
                       'number_of_companies_prov', 
                       'population_prov', 
                       'companies_prov_vs_national_%', 
                       'population_prov_vs_national_%',
                       'renta_media_prov',
                       'energetic_certif',
                       'loc_district',
                       'floor',
                       'reduced_mobility',
                       'balcony',
                       'built_in_wardrobe'], inplace=True)

### Data Deduplication
Detect and remove duplicate rows.

In [13]:
duplicated_rows = raw_data[raw_data.duplicated(keep='first')]
print(f'--------------------------------------------------')
print(f'The dataset has {duplicated_rows.shape[0]} duplicate rows')
print(f'--------------------------------------------------')

--------------------------------------------------
The dataset has 6071 duplicate rows
--------------------------------------------------


In [12]:
#Remove duplicate rows while keeping the first occurrence
raw_data.drop_duplicates(inplace=True)

#Verifying
verify_duplicated = raw_data[raw_data.duplicated(keep=False)]
print(verify_duplicated)

Empty DataFrame
Columns: [air_conditioner, bath_num, chimney, condition, construct_date, garage, garden, ground_size, heating, house_type, kitchen, lift, loc_city, loc_full, loc_neigh, loc_street, loc_zone, m2_real, m2_useful, orientation, price, room_num, storage_room, swimming_pool, terrace]
Index: []


### Handling Missing Data

In [14]:
# Missing data sorted.
missing_values = raw_data.isnull().sum().sort_values(ascending=False)
print(missing_values[missing_values > 0])

ground_size       93928
kitchen           91902
loc_street        80147
heating           69857
construct_date    63150
orientation       56789
garage            56130
loc_neigh         52086
m2_useful         44501
lift              38965
condition         13295
dtype: int64


Many variables in the dataset contain a significant amount of missing data. Upon closer investigation, it is possible that these are **structural missing values** in categorical variables and should be replaced with 0. This adjustment will be further analyzed in the following sections.

Other variables with a large amount of missing data **(more than 60%)** that should have values in all records will be removed from the dataset.

In [15]:
raw_data.drop(columns=['ground_size', 
                       'kitchen', 
                       'loc_street', 
                       'heating',
                       'construct_date',
                       'orientation',
                       'loc_neigh',
                       'm2_useful',
                       'lift'], inplace=True)

### Individual Variable Analysis for Data Cleaning

#### condition

In [16]:
raw_data['condition'].value_counts()

condition
segunda mano/buen estado      67899
segunda mano/para reformar    10457
promoción de obra nueva        2277
condition                         1
Name: count, dtype: int64

In [17]:
raw_data[raw_data['condition'] == 'condition']

air_conditioner  bath_num  chimney  condition  garage  garden  \
4248  air_conditioner  bath_num  chimney  condition  garage  garden   

      house_type  loc_city  loc_full  loc_zone  m2_real  price  room_num  \
4248  house_type  loc_city  loc_full  loc_zone  m2_real  price  room_num   

      storage_room  swimming_pool  terrace  
4248  storage_room  swimming_pool  terrace

A value equal to the header was detected, likely due to a row left over after removing duplicates. This row probably originates from instances where column names were repeated at regular intervals in the dataframe. It will be removed.

In [18]:
raw_data.drop(raw_data[raw_data['condition'] == 'condition'].index, inplace=True)

With 13,319 missing records in this variable, we deduce that the data were not recorded because they were irrelevant to buyers. This suggests that the properties are neither newly constructed nor in need of renovation, so it can be inferred that they are second-hand homes. The values will be renamed according to the terminology used in the U.S. real estate market.

In [19]:
raw_data['condition'] = raw_data['condition'].fillna('Resale')
raw_data['condition'] = raw_data['condition'].replace({
    'segunda mano/buen estado': 'Resale', 
    'segunda mano/para reformar': 'Fixer-upper', 
    'promoción de obra nueva':'New'
    })

In [20]:
# Verifying
raw_data['condition'].unique()

array(['Resale', 'Fixer-upper', 'New'], dtype=object)

#### garage

In [21]:
unique_garage = raw_data['garage'].unique()
unique_garage[:5]

array(['plaza de garaje incluida en el precio', nan,
       'plaza de garaje por 15.000 eur adicionales',
       'plaza de garaje por 12.500 eur adicionales',
       'plaza de garaje por 23.000 eur adicionales'], dtype=object)

The variable contains 3 types of values:
- 'NaN'
- 'plaza de garaje incluida en el precio'
- 'plaza de garaje por "x" eur adicionales'

The variable values are changed to:
- NaN to `Not Included`.
- 'plaza de garaje incluida en el precio' to `Included`.
- All occurrences of 'plaza de garaje por...' to `Separate Purchase`.

In [22]:
raw_data['garage'] = raw_data['garage'].fillna('Not Included')
raw_data['garage'] = raw_data['garage'].replace('plaza de garaje incluida en el precio', 'Included')
raw_data['garage'] = raw_data['garage'].replace(to_replace = r'^plaza de garaje por.*', value='Separate Purchase', regex=True)

In [23]:
# Verifiying ..
raw_data['garage'].unique()

array(['Included', 'Not Included', 'Separate Purchase'], dtype=object)

#### Verify the missing values

In [24]:
missing_values = raw_data.isnull().sum().sort_values(ascending=False)
print(missing_values)

air_conditioner    0
bath_num           0
chimney            0
condition          0
garage             0
garden             0
house_type         0
loc_city           0
loc_full           0
loc_zone           0
m2_real            0
price              0
room_num           0
storage_room       0
swimming_pool      0
terrace            0
dtype: int64


#### air_conditioner

In [25]:
raw_data['air_conditioner'].value_counts()

air_conditioner
0    71158
1    22770
Name: count, dtype: int64

#### swimming_pool

In [26]:
raw_data['swimming_pool'].value_counts()

swimming_pool
0    76108
1    17820
Name: count, dtype: int64

#### storage_room

In [27]:
raw_data['storage_room'].value_counts()

storage_room
0    55859
1    38069
Name: count, dtype: int64

#### garden

In [28]:
raw_data['garden'].value_counts()

garden
0    67872
1    26056
Name: count, dtype: int64

#### terrace

In [29]:
raw_data['terrace'].value_counts()

terrace
0    48903
1    45025
Name: count, dtype: int64

#### room_num

In [30]:
raw_data['room_num'].unique()

array(['4', '3', '2', '7', '5', '1', '8', '13', 'sin habitación', '6',
       '9', '10', '15', '12', '30', '18', '35', '17', '14', '11', '16',
       '36', '84', '20', '22', '65', '24', '23', '32', '25', '34', '26',
       '56', '58', '73'], dtype=object)

In [31]:
# Replace the value 'sin habitacion' with '0'
raw_data['room_num'] = raw_data['room_num'].replace('sin habitación', '0')

In [32]:
# Convert the variable to the correct data type
raw_data['room_num'] = pd.to_numeric(raw_data['room_num'])
# Verifying
raw_data['room_num'].dtype

dtype('int64')

#### price

In [33]:
# Convert the variable to the correct data type
raw_data['price'] = raw_data['price'].astype(float)
raw_data['price'].dtype

dtype('float64')

#### m2_real

In [34]:
# Convert the variable to the correct data type
raw_data['m2_real'] = pd.to_numeric(raw_data['m2_real'])
# verifying
raw_data['m2_real'].dtype

dtype('int64')

#### house_type

In [35]:
raw_data['house_type'].value_counts()

house_type
Piso                                       50206
 Casa o chalet independiente               13032
 Casa o chalet                              6411
 Chalet adosado                             5937
Ático                                       3261
 Casa de pueblo                             2977
Dúplex                                      2384
 Chalet pareado                             2304
 Finca rústica                              2254
Alquiler de Piso                            2232
 Casa rural                                  925
Estudio                                      678
Casa rural                                   571
 Caserón                                     147
Alquiler de Ático                            101
 Masía                                       100
Alquiler de Dúplex                            72
 Casa terrera                                 68
Alquiler de Casa o chalet independiente       61
Alquiler de Chalet adosado                    54
 Cortijo 

Remove records referring to rentals

In [36]:
pattern = r'^Alquiler.*'
alquiler_rows = raw_data[raw_data['house_type'].str.contains(pattern, regex=True)]
raw_data.drop(alquiler_rows.index, inplace=True)

In [37]:
# Remove spaces before and after strings
raw_data['house_type'] = raw_data['house_type'].str.strip()
# verifying
raw_data['house_type'].value_counts()

house_type
Piso                           50206
Casa o chalet independiente    13032
Casa o chalet                   6411
Chalet adosado                  5937
Ático                           3261
Casa de pueblo                  2977
Dúplex                          2384
Chalet pareado                  2304
Finca rústica                   2254
Casa rural                      1496
Estudio                          678
Caserón                          147
Masía                            100
Casa terrera                      68
Cortijo                           30
Palacio                           18
Torre                             18
Castillo                           8
Name: count, dtype: int64

Change values to U.S. real estate market terminology:

- **Piso**: Apartment
- **Casa o chalet independiente**: Single-Family Home
- **Casa o chalet**: Single-Family Home
- **Chalet adosado**: Townhouse
- **Ático** : Penthouse
- **Casa de pueblo**: Cottage
- **Dúplex**: Duplex
- **Chalet pareado**: Semi-Detached House
- **Finca rústica**: Farmhouse
- **Casa rural**: Rural House
- **Estudio**: Studio Apartment
- **Caserón**: Mansion
- **Masía**: Country House
- **Casa terrera**: Bungalow
- **Cortijo**: Ranch House
- **Torre**: Tower House
- **Palacio** : Palace
- **Castillo** : Castle

In [38]:
house_type_dict = {
    'Piso': 'Apartment',
    'Casa o chalet independiente': 'Single-Family Home',
    'Casa o chalet': 'Single-Family Home',
    'Chalet adosado': 'Townhouse',
    'Ático': 'Apartment',
    'Casa de pueblo': 'Cottage',
    'Dúplex': 'Duplex',
    'Chalet pareado': 'Semi-Detached House',
    'Finca rústica': 'Farmhouse',
    'Casa rural': 'Rural House',
    'Estudio': 'Studio Apartment',
    'Caserón': 'Mansion',
    'Masía': 'Country House',
    'Casa terrera': 'Bungalow',
    'Cortijo': 'Ranch House',
    'Torre': 'Tower House',
    'Palacio': 'Palace',
    'Castillo': 'Castle'
}
raw_data['house_type'] = raw_data['house_type'].map(house_type_dict)
# Verifying
raw_data['house_type'].value_counts()

house_type
Apartment              53467
Single-Family Home     19443
Townhouse               5937
Cottage                 2977
Duplex                  2384
Semi-Detached House     2304
Farmhouse               2254
Rural House             1496
Studio Apartment         678
Mansion                  147
Country House            100
Bungalow                  68
Ranch House               30
Palace                    18
Tower House               18
Castle                     8
Name: count, dtype: int64

#### bath_num

In [39]:
raw_data['bath_num'].value_counts().sort_values(ascending=False)

bath_num
2            37095
1            34080
3            12044
4             4559
5             1661
6              690
sin baños      635
7              255
8              115
9               60
10              43
11              23
13              17
12              14
14              10
15               9
20               4
18               3
16               2
17               2
34               1
84               1
19               1
23               1
32               1
30               1
58               1
75               1
Name: count, dtype: int64

In [40]:
raw_data['bath_num'] = raw_data['bath_num'].replace('sin baños', '0')
# Verifying
raw_data['bath_num'].unique()

array(['2', '3', '1', '4', '7', '6', '15', '5', '0', '8', '10', '34',
       '11', '9', '14', '12', '13', '20', '84', '16', '19', '23', '32',
       '18', '17', '30', '58', '75'], dtype=object)

#### loc_full

In [41]:
raw_data['loc_full'].head()

0              La iglesia , Urcabustaiz , Zuya, Álava 
1     la Union Plazatxoa, 8 , Legutiano , Zuya, Álava 
2          Calle Etxaguen s/n , Aramaio , Zuya, Álava 
3       CP 01449, 1449 CP , Urcabustaiz , Zuya, Álava 
4     Calle Nueva Plaza, 9 , Urcabustaiz , Zuya, Ál...
Name: loc_full, dtype: object

Remove the word 'Calle' to normalize addresses in preparation for geocoding

In [42]:
raw_data['loc_full'] = raw_data['loc_full'].str.replace(r'\b[Cc]alle\b', '', regex=True)
# Verifiying
raw_data['loc_full'].head()

0              La iglesia , Urcabustaiz , Zuya, Álava 
1     la Union Plazatxoa, 8 , Legutiano , Zuya, Álava 
2                Etxaguen s/n , Aramaio , Zuya, Álava 
3       CP 01449, 1449 CP , Urcabustaiz , Zuya, Álava 
4          Nueva Plaza, 9 , Urcabustaiz , Zuya, Álava 
Name: loc_full, dtype: object

### Data derivation

To obtain coordinates and a well-constructed complete address, the geolocator requires the street, number, and city. If any information is inaccurate, it returns `None`. In cases where a record contains incorrect data and cannot be geolocated, we will attempt to use fewer details, such as the city and municipality. The geolocator will then return the coordinates of the city. Examples:

The incorrect initial content is removed, leaving the remaining data (city and province):
- **CP 01449, 1449 CP, Urcabustaiz, Zuya, Álava**: `None`
- **Urcabustaiz, Zuya, Álava**: Urkabustaiz, Gorbeialdea, Araba/Álava, Euskadi, Spain, (42.9709057, -2.9186291298186964).

For this case, we create a new column by combining the `loc_city` and `loc_zone` variables to use this data when it is not possible to geolocate using the `loc_full` variable.

In [43]:
raw_data['city_prov'] = raw_data['loc_city'].str.cat(raw_data['loc_zone'], sep=', ')
raw_data['city_prov'].head()

0    Urcabustaiz, Zuya, Álava
1      Legutiano, Zuya, Álava
2        Aramaio, Zuya, Álava
3    Urcabustaiz, Zuya, Álava
4    Urcabustaiz, Zuya, Álava
Name: city_prov, dtype: object

Remove the `loc_city` and `loc_zone` variables from the dataframe.

In [44]:
raw_data = raw_data.drop(columns=['loc_city', 'loc_zone'])

### Saving the Cleaned Data

#### Validating Data Before Saving

In [45]:
raw_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 91329 entries, 0 to 99999
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   air_conditioner  91329 non-null  object 
 1   bath_num         91329 non-null  object 
 2   chimney          91329 non-null  object 
 3   condition        91329 non-null  object 
 4   garage           91329 non-null  object 
 5   garden           91329 non-null  object 
 6   house_type       91329 non-null  object 
 7   loc_full         91329 non-null  object 
 8   m2_real          91329 non-null  int64  
 9   price            91329 non-null  float64
 10  room_num         91329 non-null  int64  
 11  storage_room     91329 non-null  object 
 12  swimming_pool    91329 non-null  object 
 13  terrace          91329 non-null  object 
 14  city_prov        91329 non-null  object 
dtypes: float64(1), int64(2), object(12)
memory usage: 11.1+ MB


Some numerical variables were converted to object type due to mixed data types within the same column and how the data was loaded from the CSV file. The following function corrects this issue and assigns the correct data type to each column.

In [46]:
# List of columns to convert from object to numeric if possible
columns_to_convert = ['air_conditioner', 'bath_num', 'chimney', 'garden',
                      'storage_room', 'swimming_pool', 'terrace']

# Apply pd.to_numeric() to each column.
for column in columns_to_convert:
    raw_data[column] = pd.to_numeric(raw_data[column])

# Verifying the data types after conversion
print(raw_data.dtypes)

air_conditioner      int64
bath_num             int64
chimney              int64
condition           object
garage              object
garden               int64
house_type          object
loc_full            object
m2_real              int64
price              float64
room_num             int64
storage_room         int64
swimming_pool        int64
terrace              int64
city_prov           object
dtype: object


In [42]:
# raw_data.to_csv('data/clean_data.csv', index=False)
del raw_data
gc.collect()

30